### PR TITLE
[CI] Prune redundant CPU test overhead

### PR DESCRIPTION
**Pull request by JustinTong0323**
## Summary

- move seven source/AST ratchets out of the registered CPU suite into dependency-light `scripts/lint/` pre-commit checkers
- replace exhaustive logprob Cartesian sweeps with deterministic boundary, ordered-pair, and wider heterogeneous coverage cases
- run `cargo test --workspace` in the Rust extension cache-miss build path and include the test script in the cache key

## Measured impact

The measurements below are from the review-before-fix head `2602dcf6`, using its update-branch [Base run](https://github.com/sgl-project/sglang/actions/runs/31463438698) and [lint run](https://github.com/sgl-project/sglang/actions/runs/31463438391). The review fixes preserve the tested workload placement; fresh CI for head `fb1b5154b5` is running in [Base run #31580126621](https://github.com/sgl-project/sglang/actions/runs/31580126621).

| Metric | Control | This PR | Change |
|---|---:|---:|---:|
| Aggregate CPU runner time, adjacent unpruned control | 5,287s | 4,874s | **-413s (-7.8%)** |
| Base-a gate, equal-width 8-shard control | 13m21s | 12m23s | **-58s** |
| Directly affected work, Rust cache hit | 414s | ~70s | **~-344s (-83%)** |
| Directly affected work, Rust cache miss | 414s | ~94s | **~-320s (-77%)** |

The direct cache-hit reduction consists of:

- seven registered static ratchets: 196s in CPU workers to about 45s in lint
- Cargo workspace test: 158s on every Base-a run to zero on a cache hit; about 24s in the cache-miss build path
- two logprob tests: 60s to 25-26s

The adjacent aggregate control used nine CPU shards while this PR used eight, so aggregate runner time is the resource comparison, not a wall-clock comparison. The Base-a wall-clock comparison uses an older control with the same eight-shard width. These results are one matched control plus two cache-hit PR samples, not a post-merge p90 window.

## Coverage and correctness

The ratchets keep their existing rules and diagnostics but no longer pay registered-test subprocess and `sglang.test.test_utils` import costs. `CustomTestCase` remains unchanged for actual registered tests. The checkers and their focused unit tests live under `scripts/lint/`; pre-commit runs both the static ratchets and 11 checker tests.

The chunk-stitching coverage set drops from 2,800 sequence combinations to 84, while retaining every menu item, every ordered pair, and forward/reverse width-3 and width-4 heterogeneous cases. The fast-input set drops from 155 combinations to 40 with the same ordered-pair guarantee. The selection is deterministic rather than random sampling, and each order-sensitive menu is defined once.

The shared logprob helper is explicitly excluded from registered-test collection through one predicate used by `run_suite.py`, partition computation, and the registry checker. The existing registered `unit/entrypoints/openai/utils.py` remains collected.

The Rust workspace test is removed from Base-a but not from CI. A Rust extension cache entry can be saved only after `cargo test --workspace` succeeds. The test runs once while the Python 3.12 venv and PyO3 target fingerprint match, has a 300-second timeout, and changes to the test command invalidate existing compile-only entries.

## Review fixes

- moved all 12 pre-commit checkers and their tests from `scripts/ci/` to `scripts/lint/`, keeping the Rust workflow script under `scripts/ci/`
- added a pre-commit unittest-discovery hook and broadened static-ratchet triggers to every lint checker
- aligned Cargo test interpreter/target fingerprints and restored the 300-second timeout
- updated runtime-context and unit-test-admission guidance so static ratchets are checkers, not unittests
- shared the deterministic logprob coverage generator and documented the load-bearing menu order
- tightened the bare-`pytest.main` checker to require actual exit-code propagation
- centralized registered-test file filtering so helpers cannot be accidentally collected or real `utils.py` tests accidentally skipped

## Verification

Passed locally on head `fb1b5154b5`:

- full changed-files pre-commit suite, including all new and moved files
- 11 lint-checker unit tests
- all nine static ratchets against the full package
- registered-test validation: 1,543 collected files and 2,047 registries
- deterministic coverage assertions: exactly 84/40 unique cases with all singleton, ordered-pair, and forward/reverse cyclic cases
- full `python/` and `test/` no-bare scan
- Python compile checks, YAML parsing, shell syntax, and `git diff --check`

Fresh Base and lint CI on `fb1b5154b5` are pending. The local host does not have torch or a Rust toolchain, so the two logprob runtime tests and Cargo workspace tests rely on that CI run.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #31763861490](https://github.com/sgl-project/sglang/actions/runs/31763861490)
Latest PR Test (Extra): :x: [Run #31763860945](https://github.com/sgl-project/sglang/actions/runs/31763860945)

### CODE REVIEW

File: scripts/lint/check_decode_bookkeeping_ownership.py
Comment (hnyls2002) [inline scripts/lint/check_decode_bookkeeping_ownership.py:1]: Can we move the lint scripts to under `scripts/lint` instead of CI? The lint should also be run locally.

**Review by hnyls2002 — COMMENTED** [review]
Nice win, and the cache-key work is careful -- all three `hashFiles` sites moved together and the restore job's sparse-checkout picked up the new script, which is the easy thing to miss.

Six comments inline. The directory-placement one is the only structural ask; the rest are small.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -109,6 +109,17 @@ repos:
         language: system
         files: ^(\.github/actions/download-rust-ext/action\.yml|\.github/workflows/_pr-test-rust-ext-build\.yml)$
         pass_filenames: false
+      - id: check-no-bare-pytest-main
```
Comment (hnyls2002) [inline .pre-commit-config.yaml:112]: These 12 `check_*.py` are pre-commit entries, not CI orchestration. `scripts/ci/` otherwise holds `runner_configs.yml`, the per-platform install scripts (`amd/ cuda/ npu/ xpu/ musa/ slurm/`), `utils/compute_partitions.py`, `list_stage_models.py` -- real CI machinery. This PR takes `check_*.py` from 4 to 12, so the lint scripts now dominate a directory that isn't theirs.

Can they move to `scripts/lint/`? Keep `run_rust_workspace_tests.sh` in `scripts/ci/` though -- that one is a real workflow step, and it's baked into 3 `hashFiles` keys plus the restore job's sparse-checkout.

File: scripts/lint/test_check_no_bare_pytest_main.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import pathlib
```
Comment (hnyls2002) [inline scripts/lint/test_check_no_bare_pytest_main.py:1]: This file never runs. Zero references in the repo: not under `test/registered/` (so not CI-registered), not a pre-commit hook, and the only `unittest discover` in the repo is `ci-model-inventory.yml:50`, whose pattern is the literal filename `test_list_stage_models.py`.

Needs a hook to actually gate:

```yaml
- id: check-lint-script-tests
  name: unit tests for lint checkers
  entry: python3 -m unittest discover -s scripts/lint -p 'test_check_*.py'
  language: system
  files: ^scripts/lint/(check_|test_check_).*\.py$
  pass_filenames: false
```

File: .github/workflows/_pr-test-rust-ext-build.yml [hunk 3]
```diff
@@ -210,6 +211,7 @@ jobs:
               (cd python && SGLANG_BUILD_RUST_EXTS=all python setup.py build_rust --inplace)
               deactivate
           done
+          bash scripts/ci/run_rust_workspace_tests.sh
```
Comment (hnyls2002) [inline .github/workflows/_pr-test-rust-ext-build.yml]: This runs after `deactivate` (line 212) but still inherits `CARGO_TARGET_DIR=${cargo_target_root}/py3.12` from line 205, while `python3` on PATH is now the runner's system interpreter.

Line 184 already warns that PyO3's fingerprint tracks the interpreter. If the two differ, this re-fingerprints py3.12's target dir, so the next run's `build_rust` for 3.12 starts cold. The measured ~24s says they match on today's runner -- that's image- and `runs_on`-dependent.

```bash
CARGO_TARGET_DIR="${cargo_target_root}/test" bash scripts/ci/run_rust_workspace_tests.sh
```

Or move the call inside the loop, before the last `deactivate`.

File: scripts/ci/run_rust_workspace_tests.sh [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+cd "$(dirname "${BASH_SOURCE[0]}")/../../rust"
+cargo test --workspace
```
Comment (hnyls2002) [inline scripts/ci/run_rust_workspace_tests.sh]: The deleted test carried `BUILD_AND_RUN_TIMEOUT_S = 300`, documented so a hang fails with output rather than as an opaque job timeout. Nothing replaces it -- a hang now burns the compile job's `timeout-minutes: 60` on the single self-hosted build node that also serves sgl-kernel and docker builds.

```bash
timeout 300 cargo test --workspace
```

File: python/sglang/srt/server_args.py [hunk 5]
```diff
@@ -9450,7 +9450,7 @@ def m3_fp8_attn_gemm_enabled(args) -> bool:
 # (sglang.srt.runtime_context). The two functions below are LEGACY shims kept
 # for the existing call-sites; they publish/read the same live object by
 # reference. Do not add new call-sites — the counts are ratcheted
-# (decrease-only) by test/registered/unit/test_legacy_global_ratchet.py.
+# (decrease-only) by scripts/ci/check_legacy_global_ratchet.py.
```
Comment (hnyls2002) [inline python/sglang/srt/server_args.py]: This pointer got updated, but the docs that tell contributors where the ratchets live didn't:

```
.claude/skills/sglang-runtime-context/SKILL.md:139,195,381,388,393,405,460-462
.claude/rules/unit-test-admission.md:30
```

File: test/registered/unit/layers/test_logprob_chunk_stitching.py [hunk 6]
```diff
@@ -88,6 +88,15 @@ def get_logits_fn(states, lm_head, logits_metadata, **kwargs):
     )
 
 
+def _coverage_cases(menu, max_seqs):
```
Comment (hnyls2002) [inline test/registered/unit/layers/test_logprob_chunk_stitching.py]: `_coverage_cases` is duplicated verbatim here and in `test_logprob_fast_input.py:94`.

More importantly, its `width >= 3` arm only emits cyclically-adjacent runs over `menu`, so reordering `menu` silently changes which triples are covered and no assertion notices. Worth a shared helper plus a one-line comment on `menu` recording that its order is now load-bearing.

**Comment by JustinTong0323** [comment]
/rerun-test test/registered/sampling/test_penalty.py::TestPenalty.test_penalty_edge_cases_negative_penalty_values
